src/modules/aws_ec2/mappers/instance_block_device_mapping.ts

```diff
@@ -0,0 +1,259 @@
+    'modifyInstanceAttribute',
+    (InstanceId, BlockDeviceMappings) => ({
+      InstanceId,
+      BlockDeviceMappings,
+    }),
+  );
+
+  waitUntilAvailable(client: EC2, volumeId: string) {
+    return this.module.generalPurposeVolume.volumeWaiter(client, volumeId, (vol: AWSVolume | undefined) => {
+      // If state is not 'in-use' retry
+      if (!Object.is(vol?.State, VolumeState.AVAILABLE)) {
+        return { state: WaiterState.RETRY };
+      }
+      return { state: WaiterState.SUCCESS };
+    });
+  }
+
+  detachVolumeInternal = crudBuilder<EC2, 'detachVolume'>('detachVolume', VolumeId => ({ VolumeId }));
+
+  detachVolume = async (client: EC2, VolumeId: string) => {
+    await this.detachVolumeInternal(client, VolumeId);
+    await this.waitUntilAvailable(client, VolumeId);
+  };
+
+  waitUntilInUse(client: EC2, volumeId: string) {
+    return this.module.generalPurposeVolume.volumeWaiter(client, volumeId, (vol: AWSVolume | undefined) => {
+      // If state is not 'in-use' retry
+      if (!Object.is(vol?.State, VolumeState.IN_USE)) {
+        return { state: WaiterState.RETRY };
+      }
+      return { state: WaiterState.SUCCESS };
+    });
+  }
+
+  attachVolume = async (client: EC2, VolumeId: string, InstanceId: string, Device: string) => {
+    const result = await this.attachVolumeInternal(client, VolumeId, InstanceId, Device);
+    if (result) await this.waitUntilInUse(client, VolumeId);
+    else return false;
+    return true;
+  };
+
+  cloud: Crud<InstanceBlockDeviceMapping> = new Crud({
+    create: async (es: InstanceBlockDeviceMapping[], ctx: Context) => {
+      const out: InstanceBlockDeviceMapping[] = [];
+      for (const e of es) {
+        // if we have different regions we may be in a multi-region update situation, just skip
+        if (e.region !== e.instance.region) continue;
+
+        // if instance is not created we are in the first step, no need to create anything
+        if (!e.instance.instanceId) continue;
+        const client = (await ctx.getAwsClient(e.region)) as AWS;
+
+        // read volume details
+        if (!e.volumeId) throw new Error('Cannot attach empty volumes to an instance already created');
+        if (e.volume) {
+          if (e.volume.region !== e.instance.region)
+            throw new Error('Cannot create a mapping between different regions');
+
+          // only attach if volume is not in use. We can have the case that the volume is auto-attached on creation
+          if (e.volume.volumeId && e.volume.state === VolumeState.AVAILABLE) {
+            await this.attachVolume(client.ec2client, e.volume.volumeId, e.instance.instanceId, e.deviceName);
+          } else continue;
+          out.push(e);
+        }
+      }
+      return out;
+    },
+    read: async (ctx: Context, id?: string) => {
+      if (id) {
+        // decompose the id
+        const { instanceId, volumeId, region } = this.idFields(id);
```

Comment:
are we sure this is returning the right values? I'm asking since we have the mix between overrides and `@cloudId`s

Reply:
i need to have it here, as the fields for id are complex.. i actually tried to get rid of @cloudId but the tooling doesn't let me do it, it complains that if i want to map an entity, i need to have at least one cloudId.

Reply:
You can get rid of the `@cloudId` decorators if you define the `entityId` function yourself, you have the `CodedeployDeploymentGroup` case you already worked on, for example. The only thing is that if you remove the `@cloudId`s you have to write yourself the `entityId`, `idFields` and `generateId` methods in the mapper

Reply:
i added but still getting errors